File: software/examples/c/example_interrupt.c
#include <stdio.h>

#include "ip_connection.h"
#include "bricklet_io16.h"

#define HOST "localhost"
#define PORT 4223
#define UID "XYZ" // Change to your UID

// Callback function for interrupts
void cb_interrupt(char port, uint8_t interrupt_mask, uint8_t value_mask, void *user_data) {
	(void)user_data; // avoid unused parameter warning

	printf("Interrupt on port: %c\n", port);
	printf("Interrupt by: %d\n", interrupt_mask);
	printf("Value: %d\n", value_mask);
}

int main(void) {
	// Create IP connection
	IPConnection ipcon;
	ipcon_create(&ipcon);

	// Create device object
	IO16 io16;
	io16_create(&io16, UID, &ipcon);

	// Connect to brickd
	if(ipcon_connect(&ipcon, HOST, PORT) < 0) {
		fprintf(stderr, "Could not connect\n");
		return 1;
	}
	// Don't use device before ipcon is connected

	// Register callback for interrupts
	io16_register_callback(&io16,
	                       IO16_CALLBACK_INTERRUPT,
	                       (void *)cb_interrupt,
	                       NULL);

	// Enable interrupt on pin 2 of port a
	io16_set_port_interrupt(&io16, 'a', 1 << 2);

	printf("Press key to exit\n");
	getchar();
	ipcon_destroy(&ipcon); // Calls ipcon_disconnect internally
	return 0;
}

File: software/examples/javascript/ExampleInterrupt.js
var IPConnection = require('Tinkerforge/IPConnection');
var BrickletIO16 = require('Tinkerforge/BrickletIO16');

var HOST = 'localhost';
var PORT = 4223;
var UID = 'gfp';// Change to your UID

var ipcon = new IPConnection();// Create IP connection
var io = new BrickletIO16(UID, ipcon);// Create device object

ipcon.connect(HOST, PORT,
    function(error) {
        console.log('Error: '+error);        
    }
);// Connect to brickd

// Don't use device before ipcon is connected
ipcon.on(IPConnection.CALLBACK_CONNECTED,
    function(connectReason) {
        //Enable interrupt on pin 2 of port a
        io.setPortInterrupt('a', 1 << 2);      
    }
);

// Register callback for interrupts
io.on(BrickletIO16.CALLBACK_INTERRUPT,
    // Callback function for interrupts
    function(port, interruptMask, valueMask) {
        console.log('Interrupt on port: '+port);
        console.log('Interrupt by: '+parseInt(interruptMask).toString(2));
        console.log('Value: '+parseInt(valueMask).toString(2));
        console.log();
    }
);

console.log("Press any key to exit ...");
process.stdin.on('data',
    function(data) {
        ipcon.disconnect();
        process.exit(0);
    }
);

